### PR TITLE
fix(lsp): handle reverse lookup in capabilities

### PR DESCRIPTION
**Pull request by MariaSolOs**
Addresses https://github.com/neovim/neovim/pull/27526#issuecomment-1962940151.

### CODE REVIEW

**Comment by clason** [comment]
Do we really need the reverse lookup? I kinda think that (adding reverse keys _to the same table_) was an anti-pattern...

**Comment by clason** [comment]
(Don't want to block a quick fix, but worth a brief look if avoiding adding these keys in the first place is similarly easy.)

**Comment by MariaSolOs** [comment]
> Do we really need the reverse lookup? I kinda think that (adding reverse keys _to the same table_) was an anti-pattern...

That's a good point. This was already done prior to #27526, and I kept the behavior. Seems like this has been the case since the initial design: https://github.com/neovim/neovim/pull/11336

**Comment by clason** [comment]
Oh, yes, this pattern was part of the Cambrian Explosion of Nvim 0.5 (back when men were _real_ men, and women were _real_ women, and Lua modules were _real_ Lua modules). We've revisited and rethought a few such patterns since then, and that's one I think we should reconsider as well. (Not necessarily now; it's probably a bigger refactor here.)

**Comment by lewis6991** [comment]
+2 for removing the reverse lookup.

File: runtime/lua/vim/lsp/protocol.lua [hunk 1]
```diff
@@ -723,7 +723,10 @@ function protocol.make_client_capabilities()
         codeActionLiteralSupport = {
           codeActionKind = {
             valueSet = (function()
-              local res = vim.tbl_values(protocol.CodeActionKind)
+              local res = vim.iter.filter(function(value)
```
Comment (lewis6991) [inline runtime/lua/vim/lsp/protocol.lua:726]: Why doesn't this apply to `completionItemKind` below?

**Comment by MariaSolOs** [comment]
Okay I've changed this fix to instead remove the reverse lookup.

**Comment by lewis6991** [comment]
I think there are a couple places in the codebase that rely on the lookup that need fixing up. CI will tell all.

**Comment by MariaSolOs** [comment]
> I think there are a couple places in the codebase that rely on the lookup that need fixing up. CI will tell all.

I'm also checking this manually. Moving it to a draft in the meantime.

**Comment by lewis6991** [comment]
Ref: https://github.com/neovim/neovim/pull/24564

**Comment by MariaSolOs** [comment]
Okay try 3: I want to fix https://github.com/neovim/neovim/pull/27526#issuecomment-1962940151 for now. Removing the usage of the reverse lookup should be part of a larger (future) refactor.

**Comment by clason** [comment]
Makes sense, but maybe add a todo comment to the fix about replacing it with that refactor?

**Comment by MariaSolOs** [comment]
> Makes sense, but maybe add a todo comment to the fix about replacing it with that refactor?

Isn't [this comment](https://github.com/neovim/neovim/blob/13aa962c2fcebc4bf800f626457ce6db7f74710b/runtime/lua/vim/lsp/protocol.lua#L304) enough?

**Comment by clason** [comment]
Very much so. I apologize for commenting before reviewing the code.

**Comment by MariaSolOs** [comment]
> Very much so. I apologize for commenting before reviewing the code.

No need to apologize, as always, thank you for reviewing <3

**Comment by megalithic** [comment]
💯 thanks for doing this @MariaSolOs!

**Comment by megalithic** [comment]
Good morning @MariaSolOs; just trying to decipher if another fix was already merged in for https://github.com/neovim/neovim/pull/27526#issuecomment-1962940151? 

I see that a greater refactor is wanted, so this PR (#27628) may not make it in for a while. Several of us using a the language server mentioned in the linked PR comment are stuck at a particular version of neovim until that bug is resolved.

Wanted to verify I wasn't missing a prior bug fix. :)

Thanks tons!! ❤️

**Comment by MariaSolOs** [comment]
@megalithic sorry about that. Your comment explains it well: this is the bugfix, the other PR is a larger refactor. I think this can be merged now.

@clason @lewis6991 @mfussenegger wdyt?

**Comment by lewis6991** [comment]
CI failures look meaningful.

**Comment by MariaSolOs** [comment]
> CI failures look meaningful. 

Should be okay now :)

**Comment by clason** [comment]
Lint disagrees ;)

**Comment by MariaSolOs** [comment]
> Lint disagrees ;)

Fixed (spaces are always problematic aren't they? ;) )

**Pull request merged**